File: backend/src/tripl/api/v1/audit.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query

from tripl.api.deps import SessionDep, get_owner_user
from tripl.schemas.audit import AuditEntryDetailResponse, AuditListResponse
from tripl.schemas.text_filters import FreeTextFilter
from tripl.services import audit_service

# Owner-only: this feed was the back door around two other owner-only gates.
#
# Every entry carries the request payload that produced it — since tripl-5ydt on
# ``GET /audit/{entry_id}`` alone, not on every list row — and the router had
# nothing but the shared auth dependency, so any authenticated user could read:
#
#   * ``data_source.create`` / ``.update`` payloads — the warehouse host, port,
#     database and username that data_sources.py:69 deliberately BLANKS for
#     non-owners on every direct read of the same source;
#   * ``scan_config.create`` payloads — ``base_query``, the free-text SQL that
#     authoring a scan is owner-only to protect (api/v1/scans.py:36), since it
#     reads whatever the warehouse credential can.
#
# The log is also INSTANCE-wide — ``project_slug`` is a filter, not a scope — so
# the reach was every project, not just the caller's. Passwords were never
# exposed; audit_service._redact strips them (tripl-jfm3.110).
router = APIRouter(prefix="/audit", tags=["audit"], dependencies=[Depends(get_owner_user)])


@router.get("", response_model=AuditListResponse)
async def list_audit(
    session: SessionDep,
    project_slug: Annotated[FreeTextFilter | None, Query()] = None,
    action: Annotated[FreeTextFilter | None, Query()] = None,
    user_id: Annotated[uuid.UUID | None, Query()] = None,
    # FreeTextFilter: binds into a LIKE, so a NUL aborts inside asyncpg before
    # SQL runs (tripl-8wez).
    user_email: Annotated[FreeTextFilter | None, Query()] = None,
    since: Annotated[datetime | None, Query()] = None,
    until: Annotated[datetime | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> AuditListResponse:
    return await audit_service.list_entries(
        session,
        project_slug=project_slug,
        action=action,
        user_id=user_id,
        user_email=user_email,
        since=since,
        until=until,
        limit=limit,
        offset=offset,
    )


# The payload half of an entry, which the list rows no longer carry. Owner-only
# by the router dependency above, i.e. by the same gate as the list: the fields
# quoted there (warehouse host/port/database/username, ``base_query``) live in
# the payload, so this route is the one that actually hands them out.
@router.get("/{entry_id}", response_model=AuditEntryDetailResponse)
async def get_audit_entry(
    session: SessionDep,
    entry_id: uuid.UUID,
) -> AuditEntryDetailResponse:
    entry = await audit_service.get_entry(session, entry_id)
    if entry is None:
        raise HTTPException(status_code=404, detail="Audit entry not found")
    return entry

File: frontend/src/api/audit.ts
import { api } from './client'
import type { AuditEntryDetail, AuditListResponse } from '../types'

export const auditApi = {
  list: (params?: {
    projectSlug?: string
    action?: string
    userId?: string
    userEmail?: string
    since?: string
    until?: string
    limit?: number
    offset?: number
  }) => {
    const sp = new URLSearchParams()
    if (params?.projectSlug) sp.set('project_slug', params.projectSlug)
    if (params?.action) sp.set('action', params.action)
    if (params?.userId) sp.set('user_id', params.userId)
    if (params?.userEmail) sp.set('user_email', params.userEmail)
    if (params?.since) sp.set('since', params.since)
    if (params?.until) sp.set('until', params.until)
    if (params?.limit !== undefined) sp.set('limit', String(params.limit))
    if (params?.offset !== undefined) sp.set('offset', String(params.offset))
    const qs = sp.toString()
    return api.get<AuditListResponse>(`/audit${qs ? `?${qs}` : ''}`)
  },

  // The payload of one entry. List rows carry no payload at all, so this is the
  // only way to read one — see `AuditEntry` in types/branches.ts (tripl-5ydt).
  get: (entryId: string) => api.get<AuditEntryDetail>(`/audit/${entryId}`),
}
